ios/RNDeviceInfo/RNDeviceInfo.m

```diff
@@ -598,17 +598,49 @@ - (double) getTotalDiskCapacity {
     resolve(@(self.getTotalDiskCapacity));
 }
 
-- (double) getFreeDiskStorage {
+- (double)getFreeDiskStorage {
     uint64_t freeSpace = 0;
-    NSDictionary *storage = [self getStorageDictionary];
+    NSError *error = nil;
 
-    if (storage) {
-        NSNumber *freeFileSystemSizeInBytes = [storage objectForKey: NSFileSystemFreeSize];
-        freeSpace = [freeFileSystemSizeInBytes unsignedLongLongValue];
+    // iOS 11 and above: Use NSURLVolumeAvailableCapacityForImportantUsageKey
+    // https://developer.apple.com/documentation/foundation/urlresourcekey/checking_volume_storage_capacity
+    if (@available(iOS 11.0, *)) {
+        NSURL *fileURL = [NSURL fileURLWithPath:NSHomeDirectory()];
+        NSDictionary *storageValues = [fileURL resourceValuesForKeys:@[NSURLVolumeAvailableCapacityForImportantUsageKey] error:&error];
+
+        if (error) {
+            NSLog(@"Error retrieving storage information: %@", error);
+            return 0;
+        }
+
+        if (storageValues) {
+            NSNumber *availableCapacityForImportantUsage = storageValues[NSURLVolumeAvailableCapacityForImportantUsageKey];
+            if (availableCapacityForImportantUsage) {
+                freeSpace = [availableCapacityForImportantUsage unsignedLongLongValue];
+            }
+        }
     }
-    return (double) freeSpace;
+
+    // Fallback for older iOS versions: Use NSFileSystemFreeSize
+    if (freeSpace == 0 && !error) {
```

Comment:
This control flow is confusing to me.
I understand the desire to avoid if/else and do quick-return short-circuit style conditional constructions, but this one avoids the if/else but still doesn't short-circuit, which gives me a high mental load when reading the code

I'd prefer it to be a traditional if/else where the `if(@available...` is the same above, this is now an else that isn't keyed on error existing or not or freespace already being non-zero or not

then each branch of the conditional could simply return the value it sees directly or the error directly instead of storing it in method scope

Reply:
really easy to understand now, I appreciate it